cpu/cc2538/radio/cc2538_rf_radio_ops.c

```diff
@@ -0,0 +1,471 @@
+
+    return cc2538_cca_status;
+}
+
+static int _request_cca(ieee802154_dev_t *dev)
+{
+    (void) dev;
+    if (!RFCORE->XREG_FSMSTAT1bits.RX_ACTIVE) {
+        return -EINVAL;
+    }
+
+    /* Ignore baseband processing */
+    RFCORE_XREG_RFIRQM0 &= ~RXPKTDONE;
+
+    cc2538_cca = true;
+    RFCORE_SFR_RFST = ISCLEAR;
+    RFCORE_SFR_RFST = STOP;
+    RFCORE_XREG_CSPCTRL &= ~CC2538_CSP_MCU_CTRL_MASK;
+
+    /* Execute the last program */
+    RFCORE_SFR_RFST = ISSTART;
+
+    return 0;
+}
+
+static int _set_cca_threshold(ieee802154_dev_t *dev, int8_t threshold)
+{
+    (void) dev;
+    (void) threshold;
+
+    if (threshold < CC2538_RF_SENSITIVITY) {
+        return -EINVAL;
+    }
+
+    RFCORE_XREG_CCACTRL0 &= ~CC2538_CCA_THR_MASK;
+    RFCORE_XREG_CCACTRL0 |= (threshold - CC2538_RSSI_OFFSET) & CC2538_CCA_THR_MASK;
+
+    return 0;
+}
+
+static int _config_phy(ieee802154_dev_t *dev, ieee802154_phy_conf_t *conf)
+{
+    (void) dev;
+    int8_t pow = conf->pow;
+    if (pow  < OUTPUT_POWER_MIN || pow > OUTPUT_POWER_MAX) {
+        return -EINVAL;
+    }
+
+    uint8_t channel = conf->channel;
+    cc2538_set_freq(IEEE802154_CHAN2FREQ(channel));
+    cc2538_set_tx_power(pow);
+
+    return 0;
+}
+
+static int _confirm_set_trx_state(ieee802154_dev_t *dev)
+{
+    (void) dev;
+    if (RFCORE->XREG_FSMSTAT0bits.FSM_FFCTRL_STATE == FSM_STATE_RX_CALIBRATION) {
+        return -EAGAIN;
+    }
+    return 0;
+}
+
+static int _request_set_trx_state(ieee802154_dev_t *dev, ieee802154_trx_state_t state)
+{
+
+    (void) dev;
+    bool wait_sfd = RFCORE->XREG_FSMSTAT0bits.FSM_FFCTRL_STATE > 0x2 && RFCORE->XREG_FSMSTAT0bits.FSM_FFCTRL_STATE < 0x7;
+    if ((RFCORE->XREG_FSMSTAT1bits.RX_ACTIVE && !wait_sfd) ||
+         RFCORE->XREG_FSMSTAT1bits.TX_ACTIVE) {
+        return -EBUSY;
+    }
+
+    switch(state) {
+        case IEEE802154_TRX_STATE_TRX_OFF:
+        case IEEE802154_TRX_STATE_TX_ON:
+            if (RFCORE->XREG_FSMSTAT0bits.FSM_FFCTRL_STATE != FSM_STATE_IDLE) {
+                RFCORE_SFR_RFST = ISRFOFF;
+            }
+            break;
+        case IEEE802154_TRX_STATE_RX_ON:
+            RFCORE_XREG_RFIRQM0 |= RXPKTDONE;
+            RFCORE_SFR_RFST = ISRXON;
+            break;
+    }
+
+    return 0;
+}
+
+void _irq_handler(void)
+{
+    uint_fast8_t flags_f0 = RFCORE_SFR_RFIRQF0;
+    uint_fast8_t flags_f1 = RFCORE_SFR_RFIRQF1;
+
+    RFCORE_SFR_RFIRQF0 = 0;
+    RFCORE_SFR_RFIRQF1 = 0;
+
+    if (flags_f1 & TXDONE) {
+        cc2538_rf_dev.cb(&cc2538_rf_dev, IEEE802154_RADIO_CONFIRM_TX_DONE);
+    }
+     
+    if (flags_f0 & RXPKTDONE) {
+        /* CRC check */
+        uint8_t pkt_len = rfcore_peek_rx_fifo(0);
+        if (rfcore_peek_rx_fifo(pkt_len) & CC2538_CRC_BIT_MASK) {
+            cc2538_rf_dev.cb(&cc2538_rf_dev, IEEE802154_RADIO_INDICATION_RX_DONE);
+        }
+        else {
+            /* CRC failed; discard packet */
+            RFCORE_SFR_RFST = ISFLUSHRX;
+        }
+
+    }
+
+    if (flags_f1 & CSP_STOP) {
```

Comment:
What does this mean?

Reply:
The CSP_STOP interrupt is called when the routine from the Strobe Processor finishes. This is used to read the status of the CSMA-CA operation.

Reply:
Better add that as a comment.

Reply:
done!